File: src/domain/log-watching/entities/LogEntry.ts
import { LogLevel } from '../value-objects/LogLevel';

/**
 * Represents a parsed log entry from a log file.
 */
export interface LogEntryProps {
  id: string;
  timestamp: Date | null;
  level: LogLevel;
  message: string;
  raw: string;
  lineNumber: number;
  context?: Record<string, unknown>;
  stackTrace?: string[];
}

/**
 * LogEntry entity representing a single line or multi-line log entry.
 * This is the core domain entity for log viewing.
 */
export class LogEntry {
  public readonly id: string;
  public readonly timestamp: Date | null;
  public readonly level: LogLevel;
  public readonly message: string;
  public readonly raw: string;
  public readonly lineNumber: number;
  public readonly context: Record<string, unknown>;
  public readonly stackTrace: string[];

  private constructor(props: LogEntryProps) {
    this.id = props.id;
    this.timestamp = props.timestamp;
    this.level = props.level;
    this.message = props.message;
    this.raw = props.raw;
    this.lineNumber = props.lineNumber;
    this.context = props.context ?? {};
    this.stackTrace = props.stackTrace ?? [];
    Object.freeze(this);
  }

  /**
   * Creates a new LogEntry.
   * @param props - The log entry properties
   * @returns A new LogEntry instance
   */
  static create(props: LogEntryProps): LogEntry {
    return new LogEntry(props);
  }

  /**
   * Creates a LogEntry from a raw log line with minimal parsing.
   * @param raw - The raw log line
   * @param lineNumber - The line number in the file
   * @returns A new LogEntry instance
   */
  static fromRaw(raw: string, lineNumber: number): LogEntry {
    return new LogEntry({
      id: `${lineNumber}-${Date.now()}`,
      timestamp: null,
      level: LogLevel.from('info'),
      message: raw,
      raw,
      lineNumber,
    });
  }

  /**
   * Checks if this entry has a stack trace.
   * @returns True if a stack trace is present
   */
  hasStackTrace(): boolean {
    return this.stackTrace.length > 0;
  }

  /**
   * Checks if this entry contains a URL.
   * @returns True if the message contains a URL
   */
  containsUrl(): boolean {
    return /https?:\/\/[^\s]+/.test(this.message);
  }

  /**
   * Extracts URLs from the message.
   * @returns Array of URLs found in the message
   */
  extractUrls(): string[] {
    const urlRegex = /https?:\/\/[^\s]+/g;
    return this.message.match(urlRegex) ?? [];
  }

  /**
   * Gets a display-friendly timestamp string.
   * @returns Formatted timestamp or empty string if none
   */
  get formattedTimestamp(): string {
    if (!this.timestamp) return '';
    return this.timestamp.toLocaleTimeString('en-US', {
      hour12: false,
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      fractionalSecondDigits: 3,
    });
  }
}

File: src-tauri/src/domain/log_watching/entities/log_entry.rs
//! LogEntry entity representing a parsed log line.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::domain::log_watching::LogLevel;

/// A parsed log entry from a log file.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LogEntry {
    /// Unique identifier for this entry.
    pub id: String,
    /// The parsed timestamp, if available.
    pub timestamp: Option<DateTime<Utc>>,
    /// The log severity level.
    pub level: LogLevel,
    /// The parsed message content.
    pub message: String,
    /// The raw, unparsed log line.
    pub raw: String,
    /// The line number in the source file.
    pub line_number: u64,
    /// Additional context parsed from the log entry.
    #[serde(default)]
    pub context: Option<serde_json::Value>,
    /// Stack trace lines, if present.
    #[serde(default)]
    pub stack_trace: Vec<String>,
}

impl LogEntry {
    /// Creates a new LogEntry.
    pub fn new(
        id: String,
        timestamp: Option<DateTime<Utc>>,
        level: LogLevel,
        message: String,
        raw: String,
        line_number: u64,
    ) -> Self {
        Self {
            id,
            timestamp,
            level,
            message,
            raw,
            line_number,
            context: None,
            stack_trace: Vec::new(),
        }
    }

    /// Creates a LogEntry from a raw line with minimal parsing.
    pub fn from_raw(raw: String, line_number: u64) -> Self {
        let id = format!("{}-{}", line_number, chrono::Utc::now().timestamp_millis());
        Self {
            id,
            timestamp: None,
            level: LogLevel::Info,
            message: raw.clone(),
            raw,
            line_number,
            context: None,
            stack_trace: Vec::new(),
        }
    }

    /// Checks if this entry has a stack trace.
    pub fn has_stack_trace(&self) -> bool {
        !self.stack_trace.is_empty()
    }

    /// Adds context to this entry.
    pub fn with_context(mut self, context: serde_json::Value) -> Self {
        self.context = Some(context);
        self
    }

    /// Adds a stack trace to this entry.
    pub fn with_stack_trace(mut self, stack_trace: Vec<String>) -> Self {
        self.stack_trace = stack_trace;
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_from_raw() {
        let entry = LogEntry::from_raw("Test log line".to_string(), 1);
        assert_eq!(entry.message, "Test log line");
        assert_eq!(entry.line_number, 1);
        assert_eq!(entry.level, LogLevel::Info);
    }

    #[test]
    fn test_has_stack_trace() {
        let entry = LogEntry::from_raw("Error".to_string(), 1);
        assert!(!entry.has_stack_trace());

        let entry_with_trace = entry.with_stack_trace(vec!["at file.rs:10".to_string()]);
        assert!(entry_with_trace.has_stack_trace());
    }
}
